test/zipkin/tracer_test.cc

```diff
@@ -0,0 +1,168 @@
+
+  EXPECT_NE(0ULL, root_span.traceId());           // trace id must be set
+  EXPECT_EQ(root_span.traceId(), root_span.id()); // span id and trace id must be the same
+  EXPECT_FALSE(root_span.isSet().parent_id);      // no parent set
+  EXPECT_NE(0LL, root_span.timestamp());          // span's timestamp must be set
+
+  // A CS annotation must have been added
+  EXPECT_EQ(1ULL, root_span.annotations().size());
+  Annotation ann = root_span.annotations()[0];
+  EXPECT_EQ(ZipkinCoreConstants::CLIENT_SEND, ann.value());
+  EXPECT_NE(0ULL, ann.timestamp()); // annotation's timestamp must be set
+  EXPECT_TRUE(ann.isSetEndpoint());
+  Endpoint endpoint = ann.endpoint();
+  EXPECT_EQ("127.0.0.1", endpoint.ipv4());
+  EXPECT_EQ(9000, endpoint.port());
+  EXPECT_EQ("my_service_name", endpoint.serviceName());
+  EXPECT_FALSE(endpoint.isSetIpv6());
+
+  // The tracer must have been properly set
+  EXPECT_EQ(dynamic_cast<TracerPtr>(&tracer), root_span.tracer());
+
+  // Duration is not set at span-creation time
+  EXPECT_FALSE(root_span.isSet().duration);
+
+  // ==============
+  // Test the creation of a shared-context span --> SR
+  // ==============
+
+  SpanContext root_span_context(root_span);
+  Span server_side_shared_context_span = tracer.startSpan("my_span", timestamp, root_span_context);
+
+  EXPECT_EQ(timestamp, server_side_shared_context_span.startTime());
+
+  // span name should NOT be set (it was set in the CS side)
+  EXPECT_EQ("", server_side_shared_context_span.name());
+
+  // trace id must be the same in the CS and SR sides
+  EXPECT_EQ(root_span.traceId(), server_side_shared_context_span.traceId());
+
+  // span id must be the same in the CS and SR sides
+  EXPECT_EQ(root_span.id(), server_side_shared_context_span.id());
+
+  // The parent should be the same as in the CS side (none in this case)
+  EXPECT_FALSE(server_side_shared_context_span.isSet().parent_id);
+
+  // span timestamp should not be set (it was set in the CS side)
+  EXPECT_EQ(0LL, server_side_shared_context_span.timestamp());
+  EXPECT_FALSE(server_side_shared_context_span.isSet().timestamp);
+
+  // An SR annotation must have been added
+  EXPECT_EQ(1ULL, server_side_shared_context_span.annotations().size());
+  ann = server_side_shared_context_span.annotations()[0];
+  EXPECT_EQ(ZipkinCoreConstants::SERVER_RECV, ann.value());
+  EXPECT_NE(0ULL, ann.timestamp()); // annotation's timestamp must be set
+  EXPECT_TRUE(ann.isSetEndpoint());
+  endpoint = ann.endpoint();
+  EXPECT_EQ("127.0.0.1", endpoint.ipv4());
+  EXPECT_EQ(9000, endpoint.port());
+  EXPECT_EQ("my_service_name", endpoint.serviceName());
+  EXPECT_FALSE(endpoint.isSetIpv6());
+
+  // The tracer must have been properly set
+  EXPECT_EQ(dynamic_cast<TracerPtr>(&tracer), server_side_shared_context_span.tracer());
+
+  // Duration is not set at span-creation time
+  EXPECT_FALSE(server_side_shared_context_span.isSet().duration);
+
+  // ==============
+  // Test the creation of a child span --> CS
```

Comment:
this assumes the only sort of child span is a client span. That's ok if that's the intent